betrusted-io/betrusted-ec
# Add network stack IP and DHCP support for factory OQC

This adds basic support for enough of IP, UDP, and DHCP to automatically bind an IP address after a `wlan join`. The functionality is intended to support factory OQC testing for the WF200 and antenna assembly. Many things that would be important for more general use, such as renewing the IP address lease when it expires, are still unimplemented.

This is a big pile of code. Sorry ¯\\\_(ツ)\_/¯. The commit messages are fairly detailed and probably worth reading.

Major features and changes:
- Change `opt-level` from `3` to `z` to reduce the compiled code size of the EC kernel
- Rework debug logging macros to reduce size of generated code
- Introduction of `sw/net` crate for all the network stuff that can be done under `#[forbid(unsafe_code)]`
- Extensive changes to `sw/wfx_rs/src/hal_wf200.rs` to maintain network stack state as `static mut` and deal with all the `unsafe {...}` stuff needed for compatibility with the wfx-fullmac-driver C FFI code
- Improvements to serial debug logging on UP5K UART including bigger TX buffer in the gateware and a menu of debug commands (access by typing `"AT\n"`)
- Packet filter with diagnostic counters for dropped and handled packets of several different types
- Improvements to the timekeeping mechanisms including `net::timers::Stopwatch` and `net::timers::Countdown`
- DHCP protocol state machine clocked by the EC's main event loop
- Automatic DHCP binding as part of `wlan join`
- RSSI detection (strongest response seen during `ssid scan` and signal strength of last packet when joined)
- Improved connection status string for `wlan status` response including RSSI, wf200 status, dhcp state, and ssid

To test the EC UART debug stuff:
1. Use betrusted-scripts and Raspberry Pi with JTAG cable to do the UP5K UART thing with `screen`
2. After reset, type "AT\n" to unlock the debug command menu. Case matters, but "\r" will work in place of "\n"
3. Help menu should print when debug commands are unlocked. Commands run immediately when the single character menu items are typed (typing Enter is not needed). Try the network stats command.

To test DHCP:
1. Set SSID from Xous with `wlan setssid ...` shellchat command
2. Set PSK with `wlan setpass ...`
3. Do `wlan join` and wait a couple seconds (EC UART log will show WF200 status and DHCP updates during this time)
4. Do `wlan status`
